### PR TITLE
DTLS support

### PR DESCRIPTION
**Pull request by njsmith**
Here's an initial draft of #2010. It's not done -- in particular, needs a ton more tests and docs, as well as some strategy to avoid adding a hard dependency on pyopenssl -- but the core logic is all there in some form.

It depends on upstream changes https://github.com/pyca/cryptography/pull/6138 and https://github.com/pyca/pyopenssl/pull/1026, so the tests are all going to fail for now.

### CODE REVIEW

File: trio/_dtls.py [hunk 1]
```diff
@@ -0,0 +1,959 @@
...
+            # ...or, well, OpenSSL didn't like it, so I guess we didn't.
+            return
+        old_stream = dtls._streams.get(address)
+        if old_stream is not None:
+            if old_stream._client_hello == packet:
+                # ...but it's just a duplicate of a packet we got before, so never mind.
+                return
+            else:
+                # Ok, this *really is* a new handshake; the old stream should go away.
+                old_stream._replaced()
```
Comment (gesslerpd) [inline trio/_dtls.py:628]: Got an error in this code path about bool not being callable, line 709 potentially a method/attribute name collision

**Comment by njsmith** [comment]
This is still blocked waiting on cryptography+pyopenssl to release, but I believe everything on our end is done, so if anyone is excited about reviewing, go ahead! Docs are viewable here: https://trio--2047.org.readthedocs.build/en/2047/reference-io.html#datagram-tls-support

Aside from the DTLS support itself, this also adds a MVP of a mock network layer for testing (#170). It's not exposed publicly yet, because so far it only supports UDP, the mechanism for manipulating packets in flight is awkward, and I just generally didn't want to make finalizing that API a blocker for this PR. But the DTLS tests use it extensively, and using it felt *fabulous*, so I'm excited.

This is a bit of a weird situation as far as reviewing goes, because this is a huge complex piece of work, and I got paid for it but y'all didn't. So like... any review is much appreciated, but please don't feel obliged to spend days on it or anything! It is not the end of the world if some of this code doesn't get intense scrutiny – it has an extremely thorough test suite and we can always fix stuff later.

Thanks again to [redacted] for sponsoring this work.

**Comment by njsmith** [comment]
Update: pyopenssl has pushed their release, so this no longer needs hacky workarounds to test against the dev version. And I think I tracked down the last few test failures that were happening on pypy? Including finding a fun pypy bug: https://foss.heptapod.net/pypy/pypy/-/issues/3656

Let's see what CI says.

**Comment by gesslerpd** [comment]
Seems like test coverage has gone down by a small amount causing the CI check failures.

@njsmith do you think this PR is ready for review/merge?

**Comment by pquentin** [comment]
The coverage does go down but it's not blocking in my opinion.

**Review by gesslerpd — APPROVED** [review]

**Pull request merged**